### PR TITLE
expression: rewrite builtin function: COALESCE

### PR DESCRIPTION
**Pull request by bailaohe**
This pr rewrite builtin coalesce.

### CODE REVIEW

**Comment by zz-jason** [comment]
to https://github.com/pingcap/tidb/issues/4080

File: executor/executor_test.go [hunk 1]
```diff
@@ -1858,6 +1858,26 @@ func (s *testSuite) TestMiscellaneousBuiltin(c *C) {
 	_, err = tidb.GetRows(rs)
 	c.Assert(err, NotNil)
 }
+
```
Comment (zz-jason) [inline executor/executor_test.go]: plz move these tests to `expression/integration_test.go`

File: expression/builtin_compare.go [hunk 2]
```diff
@@ -86,36 +92,181 @@ type coalesceFunctionClass struct {
 	baseFunctionClass
 }
 
-func (c *coalesceFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	if err := c.verifyArgs(args); err != nil {
+func (c *coalesceFunctionClass) getFunction(args []Expression, ctx context.Context) (sig builtinFunc, err error) {
+	// builtinCoalesce returns the first non-NULL value in the list,
+	// or NULL if there are no non-NULL values.
+	// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
```
Comment (zz-jason) [inline expression/builtin_compare.go]: move "See ..." this comment to every `builtinCoalesceXXXSig` to be compatible with our existing code.

**Comment by zz-jason** [comment]
add test in `plan/typeinfer_test.go`

File: executor/executor_test.go (hunk 1 above)
Comment (bailaohe) [inline executor/executor_test.go]: Thx, so 'place in executor_test.go' in https://pingcap.com/blog-reconstruct-built-in-function-report-zh is out of date?

Comment (zz-jason) [inline executor/executor_test.go]: yes

**Comment by bailaohe** [comment]
@zz-jason PTAL

File: expression/builtin_compare.go [hunk 3]
```diff
@@ -82,40 +88,224 @@ var (
...
+		fieldTps = append(fieldTps, arg.GetType())
+	}
+
+	// Get the aggregate field type as retType
+	retTp := types.AggFieldType(fieldTps)
+	retCTp := types.AggTypeClass(fieldTps, &retTp.Flag)
+	retEvalTp := tpString
+
+	// Infer the result evalTp
+	switch retCTp {
```
Comment (zz-jason) [inline expression/builtin_compare.go]: you can use `fieldTp2EvalTp()`, which is defined in `builtin.go`

File: util/types/field_type.go [hunk 4]
```diff
@@ -91,7 +91,8 @@ func AggTypeClass(tps []*FieldType, flag *uint) TypeClass {
 			continue
 		}
 		argTypeClass := argFieldType.ToClass()
-		if argTypeClass == ClassString && mysql.HasBinaryFlag(argFieldType.Flag) {
+		if (IsTypeBlob(argFieldType.Tp) || IsTypeVarchar(argFieldType.Tp) || IsTypeChar(argFieldType.Tp)) &&
```
Comment (zz-jason) [inline util/types/field_type.go]: why modify this ?

Comment (bailaohe) [inline util/types/field_type.go]: I found "coalesce(c_int, c_datetime)" has binary flag, which is  inconsistent with mysql output

Comment (bailaohe) [inline util/types/field_type.go]: borrowed from https://github.com/pingcap/tidb/pull/4158/commits/aa13a8fe674932ad0b9348e85010bc4e74a3f968

in pr https://github.com/pingcap/tidb/pull/4158

After it's merged I'll fix the conflicts

**Comment by bailaohe** [comment]
@zz-jason PTAL

**Comment by jackysp** [comment]
Please resolve the conflicts.

**Comment by jackysp** [comment]
Need to add json type? @hicqu

File: expression/builtin_compare_test.go [hunk 5]
```diff
@@ -111,20 +112,48 @@ func (s *testEvaluatorSuite) TestCompare(c *C) {
...
-	for _, t := range tests {
-		bf, err := funcs[ast.Coalesce].getFunction(primitiveValsToConstants(t.args), s.ctx)
-		c.Assert(err, IsNil)
-		result, err := bf.eval(nil)
-		c.Assert(result.GetValue(), Equals, t.expected)
+	for _, t := range cases {
+		f, err := newFunctionForTest(s.ctx, ast.Coalesce, primitiveValsToConstants(t.args)...)
 		c.Assert(err, IsNil)
+
+		//tp := f.GetType()
```
Comment (XuHuaiyu) [inline expression/builtin_compare_test.go]: address this  comment

File: expression/builtin_compare.go [hunk 6]
```diff
@@ -82,40 +88,208 @@ var (
...
+// See http://dev.mysql.com/doc/refman/5.7/en/comparison-operators.html#function_coalesce
+type builtinCoalesceStringSig struct {
+	baseStringBuiltinFunc
+}
+
+func (b *builtinCoalesceStringSig) evalString(row []types.Datum) (string, bool, error) {
+	sc := b.ctx.GetSessionVars().StmtCtx
+	for _, a := range b.getArgs() {
+		arg, isNull, err := a.EvalString(row, sc)
+		if err != nil {
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: can we use the following code, seems to be more elegant.
``` go
func (b *builtinCoalesceStringSig) evalString(row []types.Datum) (res string, isNull bool, err error) {
    for _, a := range b.getArgs(){
        if err !=nil || !isNull {
            break
        }
    }
    return res, isNull, errors.Trace(err)
}
```

so as the other 5 functions.

File: plan/typeinfer_test.go [hunk 7]
```diff
@@ -121,6 +121,8 @@ func (s *testPlanSuite) createTestCase4Columns() []typeInferTestCase {
 		{"c_int", mysql.TypeLong, charset.CharsetBin, mysql.BinaryFlag, 11, types.UnspecifiedLength},
 		{"c_char", mysql.TypeString, charset.CharsetUTF8, 0, 20, types.UnspecifiedLength},
 		{"c_enum", mysql.TypeEnum, charset.CharsetUTF8, 0, types.UnspecifiedLength, types.UnspecifiedLength},
+		// this doesn't pass
```
Comment (bailaohe) [inline plan/typeinfer_test.go]: refer to https://github.com/pingcap/tidb/issues/4165.
It has nothing related to this pr. I'll remove the comment.

**Comment by bailaohe** [comment]
@XuHuaiyu PTAL

File: expression/builtin_compare.go [hunk 8]
```diff
@@ -82,40 +88,190 @@ var (
...
+	if err = c.verifyArgs(args); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	fieldTps := make([]*types.FieldType, 0, len(args))
+	for _, arg := range args {
+		fieldTps = append(fieldTps, arg.GetType())
+	}
+
+	// Get the aggregate field type as retType
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: s/ Get/ Use
s/ aggregate/ aggregated
add `.` at  the end of this comment

File: expression/integration_test.go [hunk 9]
```diff
@@ -1535,6 +1535,16 @@ func (s *testIntegrationSuite) TestCompareBuiltin(c *C) {
...
 
+	// for coalesce
+	tk.MustQuery("select coalesce(NULL), coalesce(NULL, NULL), coalesce(NULL, NULL, NULL);").Check(testkit.Rows("<nil> <nil> <nil>"))
+
+	tk.MustExec("drop table if exists t2")
+	tk.MustExec("create table t2(a int, b double, c datetime, d time, e char(20), f bit(10))")
+	tk.MustExec(`insert into t2 values(1, 1.1, "2017-08-01 12:01:01", "12:01:01", "abcdef", 0b10101)`)
+
+	tk.MustQuery("select coalesce(NULL, a), coalesce(NULL, b, a), coalesce(c, NULL, a, b), coalesce(d, NULL), coalesce(d, c), coalesce(NULL, NULL, e, 1), coalesce(f), coalesce(1, a, b, c, d, e, f) from t2").
+		Check(testkit.Rows(fmt.Sprintf("1 1.1 2017-08-01 12:01:01 12:01:01 %s 12:01:01 abcdef 21 1", time.Now().Format("2006-01-02"))))
```
Comment (XuHuaiyu) [inline expression/integration_test.go]: do not code like this
use `result`

File: expression/builtin_compare.go [hunk 10]
```diff
@@ -82,40 +88,190 @@ var (
...
 	}
-	sig := &builtinCoalesceSig{newBaseBuiltinFunc(args, ctx)}
+
+	bf.tp.Flag |= retTp.Flag
+	// Set retType to BINARY(0) if all arguments are of type NULL
+	retTp.Flen, retTp.Decimal = 0, types.UnspecifiedLength
+	if retTp.Tp == mysql.TypeNull {
+		types.SetBinChsClnFlag(bf.tp)
+	} else {
+		maxIntLen := 0
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: this code is not easy to understand
may add a comment for this infering process?

**Comment by bailaohe** [comment]
@XuHuaiyu PTAL again

File: expression/builtin_compare.go [hunk 11]
```diff
@@ -82,40 +88,199 @@ var (
...
+			retTp.Flen = maxIntLen + retTp.Decimal
+			if retTp.Decimal > 0 {
+				retTp.Flen++
+			}
+			if !mysql.HasUnsignedFlag(retTp.Flag) {
+				retTp.Flen++
+			}
+			bf.tp = retTp
+		} else {
+			// Other type just set the field length to maxFlen
```
Comment (XuHuaiyu) [inline expression/builtin_compare.go]: Set the field length to maxFlen for other types.

File: expression/builtin_compare.go [hunk 12]
```diff
@@ -82,40 +88,199 @@ var (
...
+			bf.tp.Flen = maxFlen
```
Comment (bailaohe) [inline expression/builtin_compare.go]: @XuHuaiyu  I already set the field length of other types here

**Comment by shenli** [comment]
@bailaohe Please resolve the conflicts. @XuHuaiyu PTAL

**Comment by XuHuaiyu** [comment]
ci failed, ptal @bailaohe

**Comment by bailaohe** [comment]
fixed. PTAL @XuHuaiyu

File: util/types/field_type.go (hunk 4 above)
Comment (XuHuaiyu) [inline util/types/field_type.go]: #4158  which has been merged seems to not modify this?
so should we keep this unchanged?

**Comment by XuHuaiyu** [comment]
rest LGTM
PTAL  @zz-jason @jackysp

File: expression/builtin_compare.go [hunk 13]
```diff
@@ -82,40 +88,199 @@ var (
...
+			if argIntLen > maxIntLen {
+				maxIntLen = argIntLen
+			}
+			if argTp.Flen > maxFlen || argTp.Flen == types.UnspecifiedLength {
+				maxFlen = argTp.Flen
+			}
+		}
+
+		// For integer, field length = maxIntLen + (1/0 for sign bit)
+		// For decimal, field lenght = maxIntLen + maxDecimal + (1/0 for sign bit)
```
Comment (jackysp) [inline expression/builtin_compare.go]: typo `lenght`

**Pull request merged**